### PR TITLE
[WIP] Major release 1.0

### LINKED ISSUE #148: Write unit tests for version 1.0
We have a massive set of changes coming to Go backend in `features-1.0` branch. Running `make test` shows current unit test coverage. It is quite low at the moment and the goal is to bring it up to 60-80%. 

It's a good exercise for anyone willing to practice Golang programming. Don't go crazy, simplistic tests are better to maintain. We don't need all possible combinations covered, single happy flow is good for the start. Since the object model code is based on [Helm's codebase](https://github.com/helm/helm), it is expected that their practices of test fixtures might be used as example. 

If something is missing in the object model, like the ability to inject the dependencies - let's discuss and add.

The work should be split into individual PRs, using `features-1.0` as a base branch. The breakdown is:

- [ ] `api.go` - some methods are super-easy to cover, like static route serving
- [ ] `objects/kubectl.go` - looks the easiest among the logic
- [ ] `objects/releases.go` - own kind of test style 
- [ ] `objects/repos.go` -  - different kind of test style

Also, quite a bit of logic lives in `handlers`, it also needs UT coverage with fake `gin` contexts, quite doable.

### LINKED ISSUE #193: Its allowing only one namespace, where it needs to allow multiple namespaces with out cluster rol
### Description

This tied to issue 189. it now allows one namespace to set. we need to have multiple namespaces to allow with out cluster role

### Screenshots

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by undera**
There is some technical debt after initial phase of development, which we plan to address. It means massive changes to backend and frontend, thus it will require a major release. The changes will be accumulated in `features-1.0` branch and this PR.

Big changes:
- [x] Self-sufficient binary, no Helm or kubectl is needed anymore
- [ ] Migrate frontend to React

Smaller TODOs:
- [x] Namespace troubles (fixes #193)
- [x] Review the concurrency safety (e.g. repos modifying the files)
- [ ] Review caching and data reuse
- [x] Review introduced TODOs
- [x] Review logging
- [x] Review unit test coverage (closes #148)
- [x] Test all locally, including namespace variations
- [x] Test inside Docker
- [x] Test inside cluster

### CODE REVIEW

**Comment by iamvikaskumar** [comment]
Hi, I would like to contribute to the "Migrate frontend to React". Can you please help me get started, thanks.

**Comment by undera** [comment]
> Hi, I would like to contribute to the "Migrate frontend to React". Can you please help me get started, thanks.

This task will be done by @ronahk , sorry

**Comment by iamvikaskumar** [comment]
> 

sure, if there is anything else where I can contribute please let me know. I can also help @ronahk if required.

File: pkg/dashboard/handlers/helmHandlers.go [hunk 1]
```diff
@@ -183,7 +357,10 @@ func (h *HelmHandler) Tests(c *gin.Context) {
 		return
 	}
 
-	out, err := h.Data.RunTests(qp.Namespace, qp.Name)
+	panic("Needs implementing")
```
Comment (hdm23061993) [inline pkg/dashboard/handlers/helmHandlers.go:360]: @undera 
Why was this changed?
what is expected there?

Comment (undera) [reply pkg/dashboard/handlers/helmHandlers.go:360]: The whole `data.go` file is gone, the approach to interact with Helm has changed into object model. 
You need to implement the new way of test functionality. It would involve `action.NewReleaseTesting(cfg)` from Helm codebase

Comment (undera) [reply pkg/dashboard/handlers/helmHandlers.go:360]: Just look at other handlers, you'll get the idea

Comment (hdm23061993) [reply pkg/dashboard/handlers/helmHandlers.go:360]: Created PR #191

**Comment by undera** [comment]
@iamvikaskumar Actually, you can propose the React-based UI via PR, if you have time and it's not too big

**Comment by undera** [comment]
I think this effort has already produced a massive change. React migrating will probably take a lot of time and we should deliver the accumulated changes earlier.
Thus, I'm deciding to proceed with releaseing 1.0.

**Comment by hdm23061993** [comment]
@undera 
Is anything pending for Task 'Review caching and data reuse'?

**Comment by undera** [comment]
> @undera Is anything pending for Task 'Review caching and data reuse'?

There's nothing that blocks the release. In fact, the old caching functionality is not used for now, but I think we should restore a bits of it in the future. Also, I've leveraged the browser-side caching in some places.

**Comment by undera** [comment]
To summarize:
1. Self-sufficient binary, no helm/kubectl requirement
2. REST API documented
3. Ability to reconfigure chart with no access to its source
4. Allow specifying multiple working namespaces
5. Auto-update repositories when installed into cluster

**Comment by undera** [comment]
+ many smaller changes

**Comment by hdm23061993** [comment]
When do you plan to create a new release?

**Comment by undera** [comment]
I'm preparing to do it today. As soon as we hit the 3K stars mark ;)

**Comment by hdm23061993** [comment]
We already hit 3K 💯 🎉🚀

image
